keps/sig-security/2763-ambient-capabilities/README.md

```diff
@@ -0,0 +1,1076 @@
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Changes to kubernetes API (https://pkg.go.dev/k8s.io/api/core/v1)
+<<[UNRESOLVED should we reuse Add field in core.v1.SecurityContext.Capabilities]>>
+There are 2 options here:-
+- Option 1: Reuse Add field in [Capabilities](https://pkg.go.dev/k8s.io/api/core/v1#Capabilities)
+  - When a capability gets added explicitly it also gets added to the ambient set in addition to getting added to inheritable, permitted, bounding and effective sets.
+  - Default capabilities are not added to the ambient set.
+  - Simple add and drop API is easy to use from a user perspective.
+
+Option 2: Add new field to [Capabilities](https://pkg.go.dev/k8s.io/api/core/v1#Capabilities)
+  - When a capability is added using this field only then does it get added to the ambient set in addition to inheritable, permitted, bounding and effective sets.
+  - Original add behavior is untouched.
+  - Users now have 2 ways to add capabilities, but need to make sure which one works and which one doesn't.
+
+```diff
+type Capabilities struct {
+  // Added capabilities
+  // +optional
+  Add []Capability `json:"add,omitempty" protobuf:"bytes,1,rep,name=add,casttype=Capability"`
+  // Removed capabilities
+  // +optional
+  Drop []Capability `json:"drop,omitempty" protobuf:"bytes,2,rep,name=drop,casttype=Capability"`
++ // Ambient capabilities to add
++ // +optional
++ Ambient []Capability `json:"ambient,omitempty" protobuf:"bytes,3,rep,name=ambient,casttype=Capability"`
+}
+```
+
+Author proposes that we go with `Option 2` above as it leaves the existing behavior untouched and makes it clear what the utility of the new field is. It will also makes it possible to enforce which capabilities can be added using this field using PodSercurityPolicy (or replacement of PodSecurityPolicy) in the future.
+<<[/UNRESOLVED]>>
+
+`Ambient` capabilities will adhere to the following rules:
+
+1. Ambient capabilities can only be added.
+2. Default capabilities are not added to the ambient capabilities set. By default the container has an empty ambient capability set,
+3. Only capability that are explicitly added in the manifest will be added to the ambient set.
+4. Since the ambient capability set obeys the invariant that no capability can ever be ambient if it is not both permitted and inheritable, adding a capability to the ambient set will also add it to the permitted and inheritable set.
+
+#### Restricted ambient capabilities.
+
+Some capabilities like `CAP_SYS_ADMIN` and `CAP_DAC_OVERRIDE` make a non-root user very powerful and we should restrict them from being added to the ambient capabilities set.
+
+This is out of scope for alpha release.
```

Comment:
Based on what Lantao said, we should be restrictive at the beginning. So maybe we want this?

Reply:
Yeah, we can just add some api validations based on my understanding.

The alpha should be more restrictive to me. Opening more caps could be out of scope for alpha release instead.